README.md
The Plymouth Interactive Social Robots dataset
==============================================

![The dataset, visualised with one of the provided tool](doc/screenshot.png)
**Important**: *this repository **does not** contain the actual PInSoRo data
(~2TB would not work with GitHub!), only the meta data. See the [PInSoRo web
site](https://freeplay-sandbox.github.io/) for details to access the actual
data.*

This dataset contains 75 recordings of free-play interactions between 2 children
(45 recordings) or one child and one robot (30 recordings).

To learn more about the dataset, visit its [web
site](https://freeplay-sandbox.github.io/).

**You are very welcome to use this dataset for your own research. If you do so,
please cite the [The PInSoRo dataset: Supporting the data-driven study of
child-child and child-robot social
dynamics](https://doi.org/10.1371/journal.pone.0205999) article.**

Getting started
---------------

1. Download and unzip the [lastest version of the dataset from the Zenodo platform](https://doi.org/10.5281/zenodo.1043507)
2. Launch `python ./tools/visualise_dataset.py ./data/<record>/pinsoro-*.csv`.
   If everything is working as expect, this should replay the interaction as
   seen in the screenshot above.
3. Have fun! (and do not hesitate to use the source of `visualise_dataset.py` as
   a starting point for your own research)


More in depth, this repository contains two things:

- the dataset **metadata** (and **data** once you have obtained the actual data)
  in the `data/` subdirectory. [Check the README.md](data/README.md) file there
  to know more about how the data is stored.
- a few tools to explore and manipulate the dataset.
  [Check the README.md](tools/README.md) to know more.

Advanced analysis tools are available as well in [a separate
repository](https://github.com/freeplay-sandbox/analysis). Check [the dataset's
website](https://freeplay-sandbox.github.io/analysing) to know more about those.

Contact
-------

If you have specific questions about the dataset, you can contact
[Séverin
Lemaignan](mailto:severin.lemaignan@brl.ac.uk?subject="[PInSoRo]"). If your
question might be relevant to other people, please instead [open an
issue](https://github.com/freeplay-sandbox/dataset/issues).



CHANGELOG.md
Changelog
=========


18/06/2019
----------

* re-published the dataset on Zenodo, after fixing a erronous swap between motion magnitude
  and direction in all the CSV files.




tools/timeline.py
from collections import OrderedDict

CHILDCHILD = "childchild"
CHILDROBOT = "childrobot"


GOALORIENTED="goaloriented"
AIMLESS="aimless"
ADULTSEEKING="adultseeking"
NOPLAY="noplay"

TASKENGAGEMENT=(GOALORIENTED,
                AIMLESS,
                ADULTSEEKING,
                NOPLAY)

SOLITARY="solitary"
ONLOOKER="onlooker"
PARALLEL="parallel"
ASSOCIATIVE="associative"
COOPERATIVE="cooperative"

SOCIALENGAGEMENT=(SOLITARY,
                  ONLOOKER,
                  PARALLEL,
                  ASSOCIATIVE,
                  COOPERATIVE)

PROSOCIAL="prosocial"
ADVERSARIAL="adversarial"
ASSERTIVE="assertive"
FRUSTRATED="frustrated"
PASSIVE="passive"

SOCIALATTITUDE=(PROSOCIAL,
                ADVERSARIAL,
                ASSERTIVE,
                FRUSTRATED,
                PASSIVE)

CONSTRUCTS = (TASKENGAGEMENT, SOCIALENGAGEMENT, SOCIALATTITUDE)
CONSTRUCTS_NAMES = {TASKENGAGEMENT:"task_engagement",
                    SOCIALENGAGEMENT:"social_engagement",
                    SOCIALATTITUDE:"social_attitude"}

MISSINGDATA="missingdata"


class Timeline:

    def __init__(self, construct, annotations):

        self.construct = construct

        self.timeline=OrderedDict()
        self.start = 0
        self.end = 0

        self.prepare(annotations)

    def prepare(self, annotations):
        
        for a in annotations:
            for k,v in a.items():
                if k in self.construct:
                    self.timeline[v[0]] = (v[0], v[1], k)

                    if self.start == 0:
                        self.start = v[0]
                    self.end = v[1]



    def attime(self, t):
        for k, v in self.timeline.items():
            start, end, construct = v
            if t >= start and t < end:
                return construct
        return MISSINGDATA

    def __repr__(self):
        return "timeline from %f to %f (%d sec), %s" % (self.start, self.end, self.end-self.start, str(self.construct))

class TimelinePrinter:
    
    def __init__(self, id, construct, coder, coded, cdt = CHILDCHILD, p_age=0, y_age=0):
        self.id = id
        self.construct = construct
        self.tp = Timeline(construct, coded["purple"])
        self.coder = coder

        self.cdt = cdt

        # age of the purple child
        self.p_age = p_age

        self.start = self.tp.start
        self.end = self.tp.end


        if cdt == CHILDCHILD:
            self.ty = Timeline(construct, coded["yellow"])
            self.start = max(self.start, self.ty.start)
            self.end = min(self.end, self.ty.end)

            # age of the purple child
            self.y_age = y_age


    @staticmethod
    def csv_header():
        return ["id", "id_with_coder", "condition", "child", "age", "construct_class", "construct", "duration"]

    def csvrows(self):
        rows = []
        for k, v in self.tp.timeline.items():
            start, end, annotation = v
            rows.append([self.id, "%s-%s" % (self.id, self.coder), self.cdt, "purple", self.p_age, CONSTRUCTS_NAMES[self.construct], annotation, end-start])

        if self.cdt == CHILDCHILD:
            for k, v in self.ty.timeline.items():
                start, end, annotation = v
                rows.append([self.id, "%s-%s" % (self.id, self.coder), self.cdt, "yellow", self.y_age, CONSTRUCTS_NAMES[self.construct], annotation, end-start])

        return rows

 
class InterraterReliability:
    
    def __init__(self, construct, coder1, coder2, cdt = CHILDCHILD):
        self.construct = construct
        self.t1p = Timeline(construct, coder1["purple"])
        self.t2p = Timeline(construct, coder2["purple"])

        self.cdt = cdt

        # [start, end] is the shortest overlapping interval
        # between to two coders
        self.start = max(self.t1p.start, self.t2p.start)
        self.end = min(self.t1p.end, self.t2p.end)


        if cdt == CHILDCHILD:
            self.t1y = Timeline(construct, coder1["yellow"])
            self.t2y = Timeline(construct, coder2["yellow"])
            self.start = max(self.start, self.t1y.start, self.t2y.start)
            self.end = min(self.end, self.t1y.end, self.t2y.end)

        self.total_units = 0
        self.total_units_missing_data = 0
        self.total_units_agree = 0

        self.compute()

        self.percentage_agreement = self.total_units_agree * 100. / (self.total_units - self.total_units_missing_data)

        self.alpha = self.krippendorff_alpha()

    def compute(self, window = 1.):

        t = self.start

        while t < self.end:
            c1 = self.t1p.attime(t)
            c2 = self.t2p.attime(t)

            t += window
            self.total_units += 1

            if c1 == MISSINGDATA or c2 == MISSINGDATA:
                self.total_units_missing_data += 1
                continue

            if c1 == c2:
                self.total_units_agree += 1

        if self.cdt == CHILDCHILD:
            t = self.start

            while t < self.end:
                c1 = self.t1y.attime(t)
                c2 = self.t2y.attime(t)

                t += window
                self.total_units += 1

                if c1 == MISSINGDATA or c2 == MISSINGDATA:
                    self.total_units_missing_data += 1
                    continue

                if c1 == c2:
                    self.total_units_agree += 1


    def krippendorff_alpha(self, window = 1.):
        import krippendorff
        import numpy as np

        data1 = []
        data2 = []
        t = self.start

        while t < self.end:
            c1 = self.t1p.attime(t)
            if c1 == MISSINGDATA:
                data1.append(np.nan)
            else:
                data1.append(hash(c1))

            c2 = self.t2p.attime(t)
            if c2 == MISSINGDATA:
                data2.append(np.nan)
            else:
                data2.append(hash(c2))

            t += window

        return krippendorff.alpha([data1, data2], level_of_measurement='nominal')

if __name__=="__main__":

    import yaml
    import sys

    with open(sys.argv[1], 'r') as yml:
        coder1 = yaml.load(yml)

    with open(sys.argv[2], 'r') as yml:
        coder2 = yaml.load(yml)

    for construct in CONSTRUCTS:
        irr = InterraterReliability(construct, coder1, coder2)

        print("Percentage agreement: %f%% on %s" % (irr.percentage_agreement, construct))


tools/animate_poses.py
#! /usr/bin/python3

import argparse

import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.animation as animation


# Set up formatting for the movie files
Writer = animation.writers['ffmpeg']
writer = Writer(fps=30, metadata=dict(artist='Severin Lemaignan'), bitrate=1800)

import numpy as np
import pandas as pd
import transformations

START_IDX=0
SEQ_SIZE=600

WIDTH=0.960
HEIGHT=0.540

SANDTRAY_WIDTH=0.338 #m
SANDTRAY_LENGHT=0.600 #m

FPS=30.

NUM_FACIAL_LANDMARKS=70 * 2

Z_IMAGE_PLANE = 1. #m

def idot(a,b):
    return np.dot(b,a)


def update(num, data, faces, time_label, ann_labels):
    # NOTE: there is no .set_data() for 3 dim data...
    #faces[0].set_data(data.iloc[START_IDX:START_IDX+num, 7:77:2], -1*data.iloc[START_IDX:START_IDX+num, 8:78:2],0)
    #faces[0].set_3d_properties(list(range(num)))


    purple_face = np.array(
            [WIDTH*data.iloc[START_IDX+num, 10:10+NUM_FACIAL_LANDMARKS:2] - WIDTH/2, 
             HEIGHT*data.iloc[START_IDX+num, 11:11+NUM_FACIAL_LANDMARKS:2] - HEIGHT/2, 
             [Z_IMAGE_PLANE] * int(NUM_FACIAL_LANDMARKS/2), 
             [1]* int(NUM_FACIAL_LANDMARKS/2)]).transpose()

    purple_face_transformed = np.apply_along_axis(idot,1,purple_face,PURPLE_CAM_TO_CENTRE).transpose()
    #purple_face_transformed=purple_face

    faces[0][0].set_data(purple_face_transformed[0:2,:])
    faces[0][0].set_3d_properties(purple_face_transformed[2,:])

    offset = NUM_FACIAL_LANDMARKS + 69

    yellow_face = np.array(
            [WIDTH*data.iloc[START_IDX+num, 10+offset:10+offset+NUM_FACIAL_LANDMARKS:2] - WIDTH/2, 
             HEIGHT*data.iloc[START_IDX+num, 11+offset:11+offset+NUM_FACIAL_LANDMARKS:2] - HEIGHT/2, 
             [Z_IMAGE_PLANE] * int(NUM_FACIAL_LANDMARKS/2), 
             [1]* int(NUM_FACIAL_LANDMARKS/2)]).transpose()

    yellow_face_transformed = np.apply_along_axis(idot,1,yellow_face,YELLOW_CAM_TO_CENTRE).transpose()
    #yellow_face_transformed=yellow_face

    faces[1][0].set_data(yellow_face_transformed[0:2,:])
    faces[1][0].set_3d_properties(yellow_face_transformed[2,:])

    ###### 3D pose of the purple head

    px = data["purple_child_head_x"][START_IDX+num]
    py = data["purple_child_head_y"][START_IDX+num]
    pz = data["purple_child_head_z"][START_IDX+num]

    faces[2][0].set_data([px, py])
    faces[2][0].set_3d_properties(pz)

    ###### 3D pose of the yellow head
    yx = data["yellow_child_head_x"][START_IDX+num]
    yy = data["yellow_child_head_y"][START_IDX+num]
    yz = data["yellow_child_head_z"][START_IDX+num]

    faces[3][0].set_data([yx, yy])
    faces[3][0].set_3d_properties(yz)

    ###### 3D gaze
    gaze_magnitude = 0.300 #m
    pgx = data["purple_child_gaze_x"][START_IDX+num] * gaze_magnitude
    pgy = data["purple_child_gaze_y"][START_IDX+num] * gaze_magnitude
    pgz = data["purple_child_gaze_z"][START_IDX+num] * gaze_magnitude
    faces[4][0].set_data([[px,px+pgx],[py,py+pgy]])
    faces[4][0].set_3d_properties([pz,pz+pgz])

    ygx = data["yellow_child_gaze_x"][START_IDX+num] * gaze_magnitude
    ygy = data["yellow_child_gaze_y"][START_IDX+num] * gaze_magnitude
    ygz = data["yellow_child_gaze_z"][START_IDX+num] * gaze_magnitude
    faces[5][0].set_data([[yx,yx+ygx],[yy,yy+ygy]])
    faces[5][0].set_3d_properties([yz,yz+ygz])

    ####### Heads orientation
    xyz_vectors = [[0,0,0,1],[0.1,0,0,1],[0,0,0,1],[0,0.1,0,1],[0,0,0,1],[0,0,0.1,1]]

    prx = data["purple_child_head_rx"][START_IDX+num]
    pry = data["purple_child_head_ry"][START_IDX+num]
    prz = data["purple_child_head_rz"][START_IDX+num]

    yrx = data["yellow_child_head_rx"][START_IDX+num]
    yry = data["yellow_child_head_ry"][START_IDX+num]
    yrz = data["yellow_child_head_rz"][START_IDX+num]

    purple_headpose = transformations.compose_matrix(angles=[prx,pry,prz], translate=[px,py,pz])
    yellow_headpose = transformations.compose_matrix(angles=[yrx,yry,yrz], translate=[yx,yy,yz])

    purple_xyz = np.apply_along_axis(idot,1,xyz_vectors,purple_headpose)
    yellow_xyz = np.apply_along_axis(idot,1,xyz_vectors,yellow_headpose)

    vx=purple_xyz[0:2].transpose()
    faces[6][0].set_data(vx[0],vx[1])
    faces[6][0].set_3d_properties(vx[2])

    vy=purple_xyz[2:4].transpose()
    faces[7][0].set_data(vy[0],vy[1])
    faces[7][0].set_3d_properties(vy[2])

    vz=purple_xyz[4:].transpose()
    faces[8][0].set_data(vz[0],vz[1])
    faces[8][0].set_3d_properties(vz[2])

    vx=yellow_xyz[0:2].transpose()
    faces[9][0].set_data(vx[0],vx[1])
    faces[9][0].set_3d_properties(vx[2])

    vy=yellow_xyz[2:4].transpose()
    faces[10][0].set_data(vy[0],vy[1])
    faces[10][0].set_3d_properties(vy[2])

    vz=yellow_xyz[4:].transpose()
    faces[11][0].set_data(vz[0],vz[1])
    faces[11][0].set_3d_properties(vz[2])




    ################## Labels/annotations

    ann_p_label, ann_y_label = ann_labels
    if isinstance(data["annotators"][START_IDX+num], str):
        time_label.set_text("frame #%d (t=%.1fs)" % (num, num/FPS))
        ann_p_label.set_text(data.iloc[START_IDX+num,-6] + "\n" + \
                            data.iloc[START_IDX+num,-5] + "\n" + \
                            data.iloc[START_IDX+num,-4]
                            )
        if data["condition"][START_IDX+num] == "childchild":
            ann_y_label.set_text(data.iloc[START_IDX+num,-3] + "\n" + \
                                data.iloc[START_IDX+num,-2] + "\n" + \
                                data.iloc[START_IDX+num,-1])
        else:
            ann_y_label.set_text("[robot]")
    else:
        time_label.set_text("frame #%d (t=%.1fs) (no annotations)" % (num, num/FPS))
    return faces

##### Transformations of 3D points to the centre of the table

def make_transform_matrix(quaternion, translation):
    M = np.identity(4)
    T = transformations.translation_matrix(translation)
    M = np.dot(M, T)
    R = transformations.quaternion_matrix(quaternion)
    M = np.dot(M, R)

    M /= M[3, 3]

    return M
 
# Obtained with the following steps:
# $ rosparam set /use_sim_time True
# $ rosbag play --clock freeplay.bag
# $ rosrun tf static_transform_publisher -0.3 0.169 0 0 0 0 sandtray_centre sandtray 20
# $ rosrun tf tf_echo sandtray_centre camera_{purple|yellow}_rgb_optical_frame
YELLOW_CAM_TO_CENTRE_QUATERNION=[-0.530, 0.220, -0.314, 0.757]
YELLOW_CAM_TO_CENTRE_TRANSLATION=[-0.408, -0.208, 0.035]

PURPLE_CAM_TO_CENTRE_QUATERNION=[0.220, -0.530, 0.757, -0.314]
PURPLE_CAM_TO_CENTRE_TRANSLATION=[-0.408, 0.190, 0.035]

# numpy.dot({PURPLE,YELLOW}_CAM_TO_CENTRE, <vector>) transforms
# a vector <vector> from one of the camera's frame to the centre of the sandtray table frame.
YELLOW_CAM_TO_CENTRE =  make_transform_matrix(YELLOW_CAM_TO_CENTRE_QUATERNION, YELLOW_CAM_TO_CENTRE_TRANSLATION)
PURPLE_CAM_TO_CENTRE =  make_transform_matrix(PURPLE_CAM_TO_CENTRE_QUATERNION, PURPLE_CAM_TO_CENTRE_TRANSLATION)



if __name__ == "__main__":

    parser = argparse.ArgumentParser(description='PInSoRo Dataset -- Dataset player')
    parser.add_argument("-i", "--start-idx", type=int, default=0, help="Start frame")
    parser.add_argument("-l", "--length", type=int, help="# of frames to display (default: full dataset)")
    parser.add_argument("--video", nargs="?", help="If set, save the animation as a video with given filename")
    parser.add_argument("path", help="path to the dataset")

    args = parser.parse_args()

    START_IDX=args.start_idx
    SEQ_SIZE=args.length

    print("Loading dataset...")
    data=pd.read_csv(args.path)
    print("Done loading.")

    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    ax.set_proj_type('persp')
    fig.set_figwidth(30)
    fig.set_figheight(10)
    fig.set_dpi(100)


    #ax.set_aspect(1.)
    #ax.tick_params(labelbottom=False, bottom=False, left=False, labelleft=False)
    ax.set_xlim3d([-0.200, 0.600])
    ax.set_ylim3d([-0.600, 0.600])
    ax.set_zlim3d([-0.100, 0.600])
    ax.set_xlabel('X (m)')
    ax.set_ylabel('Y (m)')
    ax.set_zlabel('Z (m)')

    ax.azim=-170
    ax.elev=18

    colors = ["xkcd:royal purple", "xkcd:gold"]

    ax.plot([-SANDTRAY_LENGHT/2, -SANDTRAY_LENGHT/2, SANDTRAY_LENGHT/2, SANDTRAY_LENGHT/2, -SANDTRAY_LENGHT/2],
            [-SANDTRAY_WIDTH/2, SANDTRAY_WIDTH/2, SANDTRAY_WIDTH/2, -SANDTRAY_WIDTH/2, -SANDTRAY_WIDTH/2],
            0, color="black")

    ax.plot([0,0.100],[0,0],[0,0], color="red")
    ax.plot([0,0],[0,0.100],[0,0], color="green")
    ax.plot([0,0],[0,0],[0,0.100], color="blue")

    cw = 0.05
    ch= cw * 1. *  HEIGHT/WIDTH
    cz = Z_IMAGE_PLANE / WIDTH * (cw*2)
    camera = [[0, cw,-cw,0,cw,-cw,0,cw,cw,-cw,-cw],
              [0,ch,ch,0,-ch,-ch,0,-ch,ch,ch,-ch],
              [0, cz,cz,0,cz,cz,0,cz,cz,cz,cz],
              [1]*11]
    purple_camera = np.apply_along_axis(idot,0,camera,PURPLE_CAM_TO_CENTRE)
    yellow_camera = np.apply_along_axis(idot,0,camera,YELLOW_CAM_TO_CENTRE)

    ax.plot(*purple_camera[0:3], color="xkcd:royal purple")
    ax.plot(*yellow_camera[0:3], color="xkcd:gold")


    image_plane = [[-WIDTH/2, WIDTH/2, WIDTH/2, -WIDTH/2, -WIDTH/2],
                   [-HEIGHT/2, -HEIGHT/2, HEIGHT/2, HEIGHT/2, -HEIGHT/2],
                   [Z_IMAGE_PLANE]*5,
                   [1]*5]
    purple_image_plane = np.apply_along_axis(idot,0,image_plane,PURPLE_CAM_TO_CENTRE)
    yellow_image_plane = np.apply_along_axis(idot,0,image_plane,YELLOW_CAM_TO_CENTRE)

    ax.plot(*purple_image_plane[0:3])
    ax.plot(*yellow_image_plane[0:3])

    # 2D facial landmarks
    plots=[ax.plot([], [], [], 'o', markersize=2, color=c) for c in colors]

    # 3D heads
    plots += [ax.plot([], [], [], 'o', markersize=10, color=c) for c in colors]
    # gaze vectors
    plots += [ax.plot([], [], [], color=c) for c in colors]

    # xyz gizmo, purple head
    plots += [ax.plot([], [], [], color=c) for c in ["red","green","blue"]]
    # xyz gizmo, yellow head
    plots += [ax.plot([], [], [], color=c) for c in ["red","green","blue"]]

    time_text=ax.text(-0.500,0,0,"",None, fontsize=14)
    ann_p_text = ax.text(0, -0.300, 0, "",None, fontsize=12, color="xkcd:royal purple")
    ann_y_text = ax.text(0, 0.300,0, "",None, fontsize=12, color="xkcd:gold")

    ann_labels=[ann_p_text,ann_y_text]

    # Creating the Animation object
    line_ani = animation.FuncAnimation(fig, update, SEQ_SIZE, fargs=(data, plots, time_text, ann_labels),
                                    interval=1000/FPS, blit=False, repeat=False)

    if args.video is not None:
        print("Generating video... please be patient...")

        ax.set_aspect(540/960.)
        line_ani.save(args.video, writer=writer)
    else:
        plt.show()


analysis/annotation_plots.py
import matplotlib
import numpy as np
from matplotlib import pyplot as plt
import pandas as pd

import time

fig, (ax1, ax2, ax3) = plt.subplots(ncols=3, sharey=True)

df_annotations = pd.read_csv("annotations.csv")

plots = []

indices = {}
indices["task_engagement"] = ['goaloriented', 'aimless', 'adultseeking', 'noplay']
indices["social_engagement"] = ['solitary', 'onlooker', 'parallel', 'associative', 'cooperative']
indices["social_attitude"] = ['prosocial', 'adversarial', 'frustrated', 'assertive', 'passive']

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    df_cls = df_annotations[df_annotations["construct_class"] == cls]
    df_reshaped = df_cls.groupby(["condition", "construct"])["duration"].sum().unstack().fillna(0)
    #df_reshaped.plot(ax=ax, kind="bar", stacked=True)
    #df_reshaped.set_index("construct")
    plots.append(df_reshaped.T.reindex(index=reversed(indices[cls])).T.plot(ax=ax, kind="bar", stacked=True, legend='reverse'))

##plt.title('Title 2', fontsize=16)
#plt.xlabel('Annotated construct class', fontsize=14)
plt.yticks(np.arange(0, 50*3600, 3600*5))
#plt.ylabel('Duration', fontsize=14)
##plt.legend(['Child-child', 'Child-robot'])
#

#handle, legend = 


formatter = matplotlib.ticker.FuncFormatter(lambda s, x: "%02dh%02dm" % (s // 3600, (s % 3600) // 60))

xformatter = matplotlib.ticker.FuncFormatter(lambda s, x: "child-robot" if "robot" in s else "child-child")

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    ax.yaxis.set_major_formatter(formatter)
    ax.yaxis.grid()
    ax.set_axisbelow(True) # plot grid line below bar plots
    #ax.xaxis.set_major_formatter(xformatter)
    ax.set_xlabel("")
    for i in ax.get_xticklabels(): i.set_rotation(0)
    ax.get_legend().set_title("")
    ax.set_title(cls.replace("_", " "), fontsize=20)

    #handles, labels = ax.get_legend_handles_labels()
    #handles = [dict(zip(labels, handles))[construct] for construct in indices[cls]]
    #labels = indices[cls]
    #ax.legend(handles, labels)


plt.show()


analysis/annotation_age_construct_plot.py
import matplotlib
import numpy as np
from matplotlib import pyplot as plt
import pandas as pd
import seaborn
seaborn.set()

import time

fig, (ax1, ax2, ax3) = plt.subplots(ncols=3, sharey=True)


df_annotations = pd.read_csv("annotations.csv")

plots = []

indices = {}
indices["task_engagement"] = ['goaloriented', 'aimless', 'adultseeking', 'noplay']
indices["social_engagement"] = ['solitary', 'onlooker', 'parallel', 'associative', 'cooperative']
indices["social_attitude"] = ['prosocial', 'adversarial', 'frustrated', 'assertive', 'passive']

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    df_cls = df_annotations[df_annotations["construct_class"] == cls]
    df_sum=df_cls.groupby(["age"])["duration"].sum()
    df_sum_construct=df_cls.groupby(["construct", "age"])["duration"].sum()
    df=df_sum_construct/df_sum
    df.unstack().plot.bar(ax=ax, width=0.8)


##plt.title('Title 2', fontsize=16)
#plt.xlabel('Annotated construct class', fontsize=14)
##plt.legend(['Child-child', 'Child-robot'])
#

#handle, legend = 


formatter = matplotlib.ticker.FuncFormatter(lambda s, x: "%d%%" % int(s * 100))

xformatter = matplotlib.ticker.FuncFormatter(lambda s, x: "child-robot" if "robot" in s else "child-child")

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    ax.set_ylabel("Construct distribution per age", fontsize=20)
    #ax.set_yticks(np.arange(0, 1*60, 30*60))
    #ax.set_yticklabels(np.arange(0, 1*60, 30*60))

    ax.yaxis.set_major_formatter(formatter)
    ax.yaxis.grid()
    ax.set_axisbelow(True) # plot grid line below bar plots
    #ax.xaxis.set_major_formatter(xformatter)
    ax.set_xlabel("")
    for i in ax.get_xticklabels(): i.set_rotation(30)
    ax.get_legend().set_title("Age")
    ax.set_title(cls.replace("_", " "), fontsize=20)

    #handles, labels = ax.get_legend_handles_labels()
    #handles = [dict(zip(labels, handles))[construct] for construct in indices[cls]]
    #labels = indices[cls]
    #ax.legend(handles, labels)


plt.show()


analysis/annotation_mean_std_plot.py
import matplotlib
import numpy as np
from matplotlib import pyplot as plt
import pandas as pd
import seaborn
seaborn.set()


import time

fig, (ax1, ax2, ax3) = plt.subplots(ncols=3, sharey=True)


df_annotations = pd.read_csv("annotations.csv")

plots = []

indices = {}
indices["task_engagement"] = ['goaloriented', 'aimless', 'adultseeking', 'noplay']
indices["social_engagement"] = ['solitary', 'onlooker', 'parallel', 'associative', 'cooperative']
indices["social_attitude"] = ['prosocial', 'adversarial', 'frustrated', 'assertive', 'passive']

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    df_cls = df_annotations[df_annotations["construct_class"] == cls]
    #df_sum_by_id=df_cls.groupby(["construct", "id_with_coder"]).sum().unstack(fill_value=0)

    #colors = ["blue", "orange"]
    #positions = [0,1,2,3,4,5,6]

    df=df_cls.groupby(["construct", "condition", "id_with_coder"])["duration"].sum().unstack(fill_value=0).T

    df.mean().unstack(fill_value=0).plot(kind="bar", width=0.8, yerr=df.std().unstack(), ax=ax, capsize=2)


##plt.title('Title 2', fontsize=16)
#plt.xlabel('Annotated construct class', fontsize=14)
##plt.legend(['Child-child', 'Child-robot'])
#

#handle, legend = 


formatter = matplotlib.ticker.FuncFormatter(lambda s, x: "%02dm%02ds" % (s // 60, s % 60))

xformatter = matplotlib.ticker.FuncFormatter(lambda s, x: "child-robot" if "robot" in s else "child-child")

for cls, ax in zip(["task_engagement", "social_engagement", "social_attitude"], (ax1, ax2, ax3)):
    ax.set_ylabel("Average time per annotation", fontsize=20)
    #ax.set_yticks(np.arange(0, 1*60, 30*60))
    #ax.set_yticklabels(np.arange(0, 1*60, 30*60))

    ax.yaxis.set_major_formatter(formatter)
    ax.yaxis.grid()
    ax.set_axisbelow(True) # plot grid line below bar plots
    #ax.xaxis.set_major_formatter(xformatter)
    ax.set_xlabel("")
    for i in ax.get_xticklabels(): i.set_rotation(30)
    ax.get_legend().set_title("")
    ax.set_title(cls.replace("_", " "), fontsize=20)

    #handles, labels = ax.get_legend_handles_labels()
    #handles = [dict(zip(labels, handles))[construct] for construct in indices[cls]]
    #labels = indices[cls]
    #ax.legend(handles, labels)


plt.show()
